### PR TITLE
Objects4

### PR DESCRIPTION
**Pull request by louist103**
Decompile:
object_dog
object_kz
object_link_boy

Fix issues in object_link_child

All 3 objects are fully decompiled.  I will do a larger cleanup pass on the xml (node placements and names and such)
I also fixed the limbs in object_link_child and added some missing symbols. Both objects should now be fully shiftable.

### CODE REVIEW

File: assets/xml/objects/object_dog.xml [hunk 1]
```diff
@@ -1,41 +1,41 @@
...
-        <Limb Name="object_dog_Limb_007248" LimbType="Standard" Offset="0x7248"/>
-        <Limb Name="object_dog_Limb_007254" LimbType="Standard" Offset="0x7254"/>
-        <Skeleton Name="object_dog_Skel_007290" Type="Flex" LimbType="Standard" Offset="0x7290"/>
+        <Animation Name="gDogBarkAnim" Offset="0x278" />
+        <Animation Name="gDogBowAnim" Offset="0x928" />
+        <Animation Name="gDogBow2Anim" Offset="0xC28" />
+        <Animation Name="gDogRunAnim" Offset="0xD78" />
+        <Animation Name="gDogSitAnim" Offset="0x1150" />
+        <Animation Name="gDogWalkAnim" Offset="0x1368" />
+        <Texture Name="gDogFur1Tex" OutName="tex_001380" Format="rgba16" Width="32" Height="32" Offset="0x1380" /> <!-- Lower body-->
```
Comment (Dragorn421) [inline assets/xml/objects/object_dog.xml:9]:
```suggestion
        <Texture Name="gDogFur1Tex" OutName="fur_1" Format="rgba16" Width="32" Height="32" Offset="0x1380" /> <!-- Lower body-->
```

File: assets/xml/objects/object_dog.xml [hunk 2]
```diff
@@ -1,41 +1,41 @@
...
+        <Texture Name="gDogFur2Tex" OutName="fur_2" Format="rgba16" Width="32" Height="32" Offset="0x1B80" /> <!-- Front of ears, upper part of the tail, and around the feet-->
+        <Texture Name="gDogFur3Tex" OutName="tex_002380" Format="rgba16" Width="32" Height="32" Offset="0x2380" /> <!-- Upper body, top of head-->
```
Comment (Dragorn421) [inline assets/xml/objects/object_dog.xml:11]:
```suggestion
        <Texture Name="gDogFur3Tex" OutName="fur_3" Format="rgba16" Width="32" Height="32" Offset="0x2380" /> <!-- Upper body, top of head-->
```

File: assets/xml/objects/object_dog.xml [hunk 3]
```diff
@@ -1,41 +1,41 @@
...
+        <Texture Name="gDogFaceEyeTex" OutName="face_and_eye" Format="rgba16" Width="32" Height="32" Offset="0x2B80" />
+        <Texture Name="gDogFaceNoseTex" OutName="face_and_nose" Format="rgba16" Width="32" Height="32" Offset="0x3380" />
+        <Texture Name="gDogFur4Tex" OutName="fur_3" Format="rgba16" Width="32" Height="32" Offset="0x3B80" /> <!-- Back of ears, side of head, and part of front body-->
```
Comment (Dragorn421) [inline assets/xml/objects/object_dog.xml:14]:
```suggestion
        <Texture Name="gDogFur4Tex" OutName="fur_4" Format="rgba16" Width="32" Height="32" Offset="0x3B80" /> <!-- Back of ears, side of head, and part of front body-->
```

File: assets/xml/objects/object_dog.xml [hunk 4]
```diff
@@ -1,41 +1,41 @@
...
-        <Limb Name="object_dog_Limb_007200" LimbType="Standard" Offset="0x7200"/>
-        <Limb Name="object_dog_Limb_00720C" LimbType="Standard" Offset="0x720C"/>
-        <Limb Name="object_dog_Limb_007218" LimbType="Standard" Offset="0x7218"/>
-        <Limb Name="object_dog_Limb_007224" LimbType="Standard" Offset="0x7224"/>
-        <Limb Name="object_dog_Limb_007230" LimbType="Standard" Offset="0x7230"/>
-        <Limb Name="object_dog_Limb_00723C" LimbType="Standard" Offset="0x723C"/>
-        <Limb Name="object_dog_Limb_007248" LimbType="Standard" Offset="0x7248"/>
-        <Limb Name="object_dog_Limb_007254" LimbType="Standard" Offset="0x7254"/>
-        <Skeleton Name="object_dog_Skel_007290" Type="Flex" LimbType="Standard" Offset="0x7290"/>
+        <Animation Name="gDogBarkAnim" Offset="0x278" />
```
Comment (Dragorn421) [inline assets/xml/objects/object_dog.xml:3]:
```suggestion
        <Animation Name="gDogBarkAnim" Offset="0x278"/>
```
some arbitrary decision I made in #1044

File: assets/xml/objects/object_kz.xml [hunk 5]
```diff
@@ -1,43 +1,45 @@
...
-        <Limb Name="object_kz_Limb_008644" LimbType="Standard" Offset="0x8644"/>
-        <Limb Name="object_kz_Limb_008650" LimbType="Standard" Offset="0x8650"/>
-        <Limb Name="object_kz_Limb_00865C" LimbType="Standard" Offset="0x865C"/>
-        <Limb Name="object_kz_Limb_008668" LimbType="Standard" Offset="0x8668"/>
+        <Animation Name="gKzMweepAnim" Offset="0x46C" />
+        <Animation Name="gKzIdleAnim" Offset="0x75C" />
+        <Texture Name="gKzTlut" OutName="tlut" Format="rgba16" Width="16" Height="16" Offset="0x770"/>
+        <Texture Name="gKzEyeTlut" OutName="eye_tlut" Format="rgba16" Width="16" Height="16" Offset="0x970"/>
+        <Texture Name="gKzFaceScalesTex" OutName="face_scales" Format="ci8" Width="32" Height="32" Offset="0xB70" TlutOffset="0x770"/>
+        <Texture Name="gKzSkinGradientTex" OutName="skin_gradient" Format="ci8" Width="8" Height="16" Offset="0xF70" TlutOffset="0x770"/>
```
Comment (Dragorn421) [inline assets/xml/objects/object_kz.xml:8]:
```suggestion
        <Texture Name="gKzSkinGradient1Tex" OutName="skin_gradient_1" Format="ci8" Width="8" Height="16" Offset="0xF70" TlutOffset="0x770"/>
```

File: assets/xml/objects/object_kz.xml [hunk 6]
```diff
@@ -1,43 +1,45 @@
...
+        <Texture Name="gKzCrownTex" OutName="crown" Format="ci8" Width="32" Height="32" Offset="0xFF0" TlutOffset="0x770"/>
+        <Texture Name="gKzSkinGradient2Tex" OutName="skin_gradient_2" Format="ci8" Width="8" Height="8" Offset="0x13F0" TlutOffset="0x770"/>
+        <Texture Name="object_kz_Tex_001430" OutName="tex_001430" Format="ci8" Width="8" Height="8" Offset="0x1430" TlutOffset="0x770"/>
+
+        <Texture Name="gKzEyeOpenTex" OutName="eye_open" Format="ci8" Width="32" Height="32" Offset="0x1470" TlutOffset="0x970"/>
+        <Texture Name="gKzEyeHalfTex" OutName="eye_half" Format="ci8" Width="32" Height="32" Offset="0x1870" TlutOffset="0x970"/>
+        <Texture Name="gKzEyeClosedTex" OutName="eye_closed" Format="ci8" Width="32" Height="32" Offset="0x1C70" TlutOffset="0x970"/>
+        <Texture Name="gKzEyeOpen2Tex" OutName="eye_open2" Format="ci8" Width="32" Height="32" Offset="0x2070" TlutOffset="0x970"/>
```
Comment (Dragorn421) [inline assets/xml/objects/object_kz.xml:16]:
```suggestion
        <Texture Name="gKzEyeOpen2Tex" OutName="eye_open_2" Format="ci8" Width="32" Height="32" Offset="0x2070" TlutOffset="0x970"/>
```

File: assets/xml/objects/object_link_boy.xml [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+
+
+        <Texture Name="gLinkAdultShieldHandleTex" OutName="shield_handle" Format="ci8" Width="8" Height="16" Offset="0xD4B8" TlutOffset="0xCD48"/>
+        <Texture Name="gLinkAdultHylianShieldBackTex" OutName="hylian_shield_back" Format="ci8" Width="16" Height="32" Offset="0xD938" TlutOffset="0xCB40"/>
+        <Texture Name="gLinkAdultSheathBandTex" OutName="sheath_band" Format="ci8" Width="32" Height="16" Offset="0xE838" TlutOffset="0xCB40"/>
+        
+        <Texture Name="gLinkAdultMirrorShieldLowerDesignTex" OutName="mirror_shield_lower_design" Format="ia16" Width="32" Height="64" Offset="0xB280"/>
+        <Texture Name="gLinkAdultMirrorShieldUpperDesignTex" OutName="mirror_shield_upper_design" Format="ia8" Width="32" Height="64" Offset="0xC280"/>
+
+        <Texture Name="gLinkAdultDefaultGauntletTex" OutName="default_gauntlet" Format="ci8" Width="16" Height="32" Offset="0xD2B8" TlutOffset="0xCD48"/>
```
Comment (Dragorn421) [inline assets/xml/objects/object_link_boy.xml:135]:
```suggestion
        <Texture Name="gLinkAdultDefaultGauntlet1Tex" OutName="default_gauntlet_1" Format="ci8" Width="16" Height="32" Offset="0xD2B8" TlutOffset="0xCD48"/>
```

File: assets/xml/objects/object_link_boy.xml [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+        <Texture Name="gLinkAdultMouth2Tex" OutName="mouth_2" Format="ci8" Width="32" Height="32" Offset="0x4400" TlutOffset="0x5C00"/>
+        <Texture Name="gLinkAdultMouth3Tex" OutName="mouth_3" Format="ci8" Width="32" Height="32" Offset="0x4800" TlutOffset="0x5C00"/>
+        <Texture Name="gLinkAdultMouth4Tex" OutName="mouth_4" Format="ci8" Width="32" Height="32" Offset="0x4C00" TlutOffset="0x5C00"/>
+        
+        <Texture Name="gLinkAdultEarTex" OutName="ear" Format="ci8" Width="16" Height="16" Offset="0x5000" TlutOffset="0x5C00"/>
+        <Texture Name="gLinkAdultNoseTex" OutName="nose" Format="ci8" Width="16" Height="16" Offset="0x5100" TlutOffset="0x5C00"/>
+        
+        <!-- Tluts -->
+        <Texture Name="gLinkAdultHeadTlut" OutName="head_tlut" Format="rgba16" Width="16" Height="16" Offset="0x5C00"/>
+        <Texture Name="gLinkBoyTlut_005E00" OutName="tlut_5E00" Format="rgba16" Width="16" Height="16" Offset="0x5E00"/>
```
Comment (Dragorn421) [inline assets/xml/objects/object_link_boy.xml:244]: there are a few occurrences of `LinkBoy` to be replaced with `LinkAdult`

File: assets/xml/objects/object_link_boy.xml [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+
+        <!-- Unused -->
+        <Texture Name="gLinkBoyUnusedHandTex" OutName="unused_hand" Format="ci8" Width="16" Height="16" Offset="0x5300" TlutOffset="0x5C00"/>
+        <Texture Name="gLinkAdultUnusedGauntletTex" OutName="unused_gauntlet" Format="ci8" Width="16" Height="32" Offset="0x7B80" TlutOffset="0x5800"/>
```
Comment (Dragorn421) [inline assets/xml/objects/object_link_boy.xml:248]:
```suggestion
        <Texture Name="gLinkAdultUnusedGauntlet1Tex" OutName="unused_gauntlet_1" Format="ci8" Width="16" Height="32" Offset="0x7B80" TlutOffset="0x5800"/>
```

File: assets/xml/objects/object_link_boy.xml [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+        <Texture Name="gLinkAdultUnusedGauntlet2Tex" OutName="unused_gauntlet_2" Format="ci8" Width="16" Height="32" Offset="0x7D80" TlutOffset="0x5800"/>
+        <Texture Name="gLinkAdultUnusedSheathTex" OutName="unused_sheath" Format="ci8" Width="32" Height="64" Offset="0x8380" TlutOffset="0x5E00"/>
+        <Texture Name="gLinkAdultUnusedSheathBandTex" OutName="unused_sheat_band" Format="ci8" Width="32" Height="16" Offset="0x8B80" TlutOffset="0x5E00"/>
+        <Texture Name="gLinkAdultUnusedSwordPommelTex" OutName="unused_sword_pommel" Format="ci8" Width="16" Height="16" Offset="0x8D80" TlutOffset="0x5E00"/>
+        <Texture Name="gLinkAdultUnusedSwordGuardTex" OutName="unused_sword_guard" Format="ci8" Width="32" Height="32" Offset="0x8E80" TlutOffset="5E00"/>
+        <Texture Name="gLinkAdultUnusedSwordEmblemTex" OutName="unused_sword_emblem" Format="ci8" Width="16" Height="16" Offset="0x9280" TlutOffset="0x5E00"/>
+
+        <Texture Name="gLinkAdultUnusedGauntletPlate1Tex" OutName="unused_gauntlet_plate_1" Format="rgba16" Width="16" Height="32" Offset="0xA180"/>
```
Comment (Dragorn421) [inline assets/xml/objects/object_link_boy.xml:256]: where is `gLinkAdultUnusedGauntletPlate2Tex`? (notice the 2)

(
```suggestion
        <Texture Name="gLinkAdultUnusedGauntletPlateTex" OutName="unused_gauntlet_plate" Format="rgba16" Width="16" Height="32" Offset="0xA180"/>
```
)

File: include/variables.h [hunk 11]
```diff
@@ -164,12 +156,10 @@ extern s16 gSpoilingItems[3];
...
 extern FlexSkeletonHeader* gPlayerSkelHeaders[2];
 extern u8 gPlayerModelTypes[][5];
-extern Gfx* D_80125D68[];
 extern Gfx* D_80125DE8[];
-extern Gfx* D_80125E08[];
+extern Gfx* gLeftHandDLs[];
 extern Gfx* D_80125E18[];
-extern Gfx* D_80125E58[];
-extern Gfx* D_80125EF8[];
+extern Gfx* gBoomerrangDLs[];
```
Comment (Dragorn421) [inline include/variables.h:162]:
```suggestion
extern Gfx* gBoomerangDLs[];
```

File: src/overlays/actors/ovl_En_Dog/z_en_dog.c [hunk 12]
```diff
@@ -78,8 +78,12 @@ typedef enum {
     /* 0x06 */ DOG_BOW_2
 } DogBehavior;
 
+typedef enum {
+     /* 0x00 */ DOG_ANIM_WALK
+} DogAnimation;
+
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Dog/z_en_dog.c:84]: what's this for?

File: assets/xml/objects/object_link_child.xml [hunk 13]
```diff
@@ -7,63 +7,63 @@
...
-        <Limb Name="gLinkChildCollarLimb" LimbType="LOD" Offset="0x2CE88"/>
-        <Limb Name="gLinkChildRightSholderLimb" LimbType="LOD" Offset="0x2CE98"/>
-        <Limb Name="gLinkChildRightForearmLimb" LimbType="LOD" Offset="0x2CEA8"/>
-        <Limb Name="gLinkChildLeftHandLimb" LimbType="LOD" Offset="0x2CEB8"/>
-        <Limb Name="gLinkChildLeftSholderLimb" LimbType="LOD" Offset="0x2CEC8"/>
-        <Limb Name="gLinkChildLeftForearmLimb" LimbType="LOD" Offset="0x2CED8"/>
-        <Limb Name="gLinkChildRightHandLimb" LimbType="LOD" Offset="0x2CEE8"/>
-        <Limb Name="gLinkChildSwordAndSheathLimb" LimbType="LOD" Offset="0x2CEF8"/>
-        <Limb Name="gLinkChildTorsoLimb" LimbType="LOD" Offset="0x2CF08"/>
+        <Limb Name="gLinkChildRootLimb" Type="LOD" Offset="0x2CDC8"/>
```
Comment (Dragorn421) [inline assets/xml/objects/object_link_child.xml:46]:
```suggestion
        <Limb Name="gLinkChildRootLimb" LimbType="LOD" Offset="0x2CDC8"/>
```
another arbitrary #1044 decision
(you may want to be more careful when merging changes)

File: assets/xml/objects/object_link_child.xml [hunk 14]
```diff
@@ -120,105 +120,105 @@
...
-        <Texture Name="gLinkChildMasterSwordGuardTex" OutName="link_child_master_sword_guard" Format="ci8" Width="32" Height="32" Offset="0xA6F0" TlutOffset="0x9F10"/>
-        <Texture Name="gLinkChildMasterSwordEmblemTex" OutName="link_child_master_sword_emblem" Format="ci8" Width="16" Height="16" Offset="0xAAF0" TlutOffset="0x9F10"/>
-        <Texture Name="gLinkChildGoronBraceletTex" OutName="link_child_goron_bracelet" Format="rgba16" Width="8" Height="8" Offset="0x7208"/>
-        <Texture Name="gLinkChildGoronSymbolTex" OutName="link_child_goron_symbol" Format="ia16" Width="16" Height="32" Offset="0x7288"/>
+        <Texture Name="gLinkChildNoseTex" OutName="nose" Format="ci8" Width="16" Height="16" Offset="0x5000" TlutOffset="0x5500"/>
+        <Texture Name="gLinkChildUnusedHandTex" OutName="unused_hand" Format="ci8" Width="16" Height="16" Offset="0x5100" TlutOffset="0x5500"/>
+        <Texture Name="gLinkChildEarTex" OutName="ear" Format="ci8" Width="16" Height="16" Offset="0x5200" TlutOffset="0x5500"/>
+        <Texture Name="gLinkChildLowerBootTex" OutName="lower_boot" Format="ci8" Height="16" Width="16" Offset="0x5C00" TlutOffset="0x5300"/>
```
Comment (Dragorn421) [inline assets/xml/objects/object_link_child.xml:126]:
```suggestion
        <Texture Name="gLinkChildLowerBootTex" OutName="lower_boot" Format="ci8" Width="16" Height="16" Offset="0x5C00" TlutOffset="0x5300"/>
```
#1044 again

**Review by Dragorn421 — CHANGES_REQUESTED** [review]

File: include/variables.h (hunk 11 above)
Comment (fig02) [inline include/variables.h:162]: we should probably prefix these arrays with player or something, seems easy to have global variable name clashes with something as generic as "left hand dlists"

File: src/code/z_player_lib.c [hunk 15]
```diff
@@ -44,197 +45,224 @@ u8 gPlayerModelTypes[][5] = {
...
 };
 
 // Indexed by model types (left hand, right hand, sheath or waist)
 Gfx** sPlayerDListGroups[] = {
-    D_80125E08, D_80125E18, D_80125E38, D_80125E28, D_80125DE8, D_80125EE8, D_80125EF8,
-    D_80125F08, D_80125E48, D_80125E58, D_80125CE8, D_80125E68, D_80125EA8, D_80125EB8,
-    D_80125EC8, D_80125ED8, D_80125E78, D_80125E88, D_80125D28, D_80125D88, D_80125E98,
+    gPlayerLeftHandDLs, D_80125E18,    D_80125E38,           D_80125E28,          D_80125DE8, sHammerDLs, gPlayerBoomerangDLs,
+    D_80125F08,   sRightHandDLs, sRightHandClosed2DLs, sRightHandClosedDLs, sHandHoldingFirstPersonWeaponDLs, D_80125EA8, sFairyOcarinaDLs,
+    sOotDLs,      sHookshotDLs,  sSwordAndSheathDLs,   sSheathDLs,          D_80125D28, D_80125D88, sWaistDLs,
```
Comment (fig02) [inline src/code/z_player_lib.c:265]: I feel like we should probably let whoever does model group documentation name these arrays. As is, I dont think I agree with some of the names here. For example the hand dlist arrays are probably going to be more appropriately named based on the model group they are indexed by/belong to rather than naming it after the fact that his hand is closed. And the bow/slingshot one being named first person weapon, when there are other first person weapons is unfortunate

Basically, I understand that it feels all over the place, and thats cause the model group system kinda is, but I think they should be documented together. I would personally just revert all of these names for now

File: src/code/z_player_lib.c [hunk 16]
```diff
@@ -44,197 +45,224 @@ u8 gPlayerModelTypes[][5] = {
...
+    gLinkAdultRightHandHoldingOotFarDL,
     gLinkChildRightHandHoldingOOTFarDL,
 };
 
-Gfx* D_80125ED8[] = {
-    0x06024D70,
+//! @bug The 'far' display list is never used
+Gfx* sHookshotDLs[] = {
+    gLinkAdultRightHandHoldingHookshotNearDL,
     gLinkChildRightHandNearDL,
```
Comment (fig02) [inline src/code/z_player_lib.c:210]: I wouldnt call this a bug, we have know way of really knowing. Its possible they ended up not wanting to use the far dlist in the end. You can remove the bug tag and say "a far dlists exists for holding hookshot, but it is not used"

File: src/code/z_player_lib.c [hunk 17]
```diff
@@ -1093,7 +1125,7 @@ void func_80090A28(Player* this, Vec3f* vecs) {
     Matrix_MultVec3f(&D_80126098, &vecs[2]);
 }
 
-void func_80090AFC(GlobalContext* globalCtx, Player* this, f32 arg2) {
+void Player_DrawHookshotDot(GlobalContext* globalCtx, Player* this, f32 arg2) {
     static Vec3f D_801260C8 = { -500.0f, -100.0f, 0.0f };
     CollisionPoly* sp9C;
```
Comment (fig02) [inline src/code/z_player_lib.c:1130]: reticle is a better term to use than dot I think. would update the dlist name as well

File: src/code/z_player_lib.c (hunk 16 above)
Comment (louist103) [reply src/code/z_player_lib.c:210]: I feel that given the fact that the child DLs have near and far it was likely intended for adult to also use near and far.

Comment (fig02) [reply src/code/z_player_lib.c:210]: likely intended sure, but its not a bug.
It was likely, in my opinon, a choice on their part, to put the near dlist as far so it would be used for both on purpose. for all we know they decided the far one didnt look good.

Comment (fig02) [reply src/code/z_player_lib.c:210]: the comment i suggested with the removal of the bug tag is more appropriate imo

Comment (fig02) [reply src/code/z_player_lib.c]: possibly put it next to where the far dlist would go. at the top its not clear what exactly its talking about

**Review by fig02 — APPROVED** [review]

**Review by Roman971 — APPROVED** [review]
